# Netflix Movie Data Analysis

This project explores a movie dataset using Python, Pandas, Matplotlib, and Seaborn.  
The analysis focuses on genres, vote categories, popularity, and release-year trends.

## Project questions

1. Which genres appear most frequently?
2. Which vote category has the highest number of movies?
3. Which movies are the most and least popular?
4. Which years had the highest number of movie releases?
5. Is there any relationship between popularity, vote count, and vote average?


## Phase-1: Importing Modules and Data Loading

In [50]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [51]:
df = pd.read_csv("mymoviedb.csv", lineterminator='\n')
df.head()

,Release_Date,Title,Overview,Popularity,Vote_Count,Vote_Average,Original_Language,Genre,Poster_Url
0,2021-12-15,Spider-Man: No Way Home,Peter Parker is unmasked and no longer able to...,5083.954,8940,8.3,en,"Action, Adventure, Science Fiction",https://image.tmdb.org/t/p/original/1g0dhYtq4i...
1,2022-03-01,The Batman,"In his second year of fighting crime, Batman u...",3827.658,1151,8.1,en,"Crime, Mystery, Thriller",https://image.tmdb.org/t/p/original/74xTEgt7R3...
2,2022-02-25,No Exit,Stranded at a rest stop in the mountains durin...,2618.087,122,6.3,en,Thriller,https://image.tmdb.org/t/p/original/vDHsLnOWKl...
3,2021-11-24,Encanto,"The tale of an extraordinary family, the Madri...",2402.201,5076,7.7,en,"Animation, Comedy, Family, Fantasy",https://image.tmdb.org/t/p/original/4j0PNHkMr5...
4,2021-12-22,The King's Man,As a collection of history's worst tyrants and...,1895.511,1793,7.0,en,"Action, Adventure, Thriller, War",https://image.tmdb.org/t/p/original/aq4Pwv5Xeu...


## Phase 2 — Initial data inspection

In [52]:
print("Shape:",df.shape)
df.info()

Shape: (9827, 9)
<class 'pandas.DataFrame'>
RangeIndex: 9827 entries, 0 to 9826
Data columns (total 9 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Release_Date       9827 non-null   str    
 1   Title              9827 non-null   str    
 2   Overview           9827 non-null   str    
 3   Popularity         9827 non-null   float64
 4   Vote_Count         9827 non-null   int64  
 5   Vote_Average       9827 non-null   float64
 6   Original_Language  9827 non-null   str    
 7   Genre              9827 non-null   str    
 8   Poster_Url         9827 non-null   str    
dtypes: float64(2), int64(1), str(6)
memory usage: 691.1 KB


In [53]:
df['Genre'].head() # in order to find the types of genre being used in the dataset

0    Action, Adventure, Science Fiction
1              Crime, Mystery, Thriller
2                              Thriller
3    Animation, Comedy, Family, Fantasy
4      Action, Adventure, Thriller, War
Name: Genre, dtype: str

In [ ]:
df.duplicated().sum()

In [ ]:
print(df.describe())

In [ ]:
# Release-Date datatype/format needs to be changed
# white spaces between the Genre types may or may not needed to be removed
# dropping down the columns that needs not analysis(like-overview, language and poster url)
# All of these things needs to be done in the data preprocessing.
# Insights gained till now-


### Exploratory summary
- we have a dataframe consisting of 9827 entries and 9 columns.
- our dataset looks a bit tidy with no NaNs or duplicated values.
- Release_Date column need to be casted into date time and to extract only the year value.
- Overview, Original_Language and Poster-Url wouldn't be so useful during analysis, so we'll drop them.
- There is noticable outliers in the Popularity column.
- Vote_Average better be categorised for proper analysis.
- Genre column has comma separated values and white spaces that needs to be handled and casted into category.

### Phase-3: Cleaning Data and Preprocessing

In [ ]:
# changing the date time format of the Release_Date column

df['Release_Date'] = pd.to_datetime(df['Release_Date'])
print(df['Release_Date'].dtypes)

df['Release_Date'] = df['Release_Date'].dt.year
df['Release_Date'].dtypes

In [ ]:
df.head()

#### Dropping down the irrelevant columns

In [ ]:
# Overview, Original_Language, Poster_Url column to be dropped

cols = ['Overview', 'Original_Language', 'Poster_Url']
df.drop(cols, axis = 1, inplace = True)
df.columns


In [ ]:
df.head()

#### **Categorizing** `Vote_Average` column  
We would cut the `Vote_Average` values and make 4 categories `popular`,`average`, `below_avg` , `not_popular` to describe it more using categorize_col() function provided

In [ ]:
# creating a user defined function to label the popularity of the movie

def categorize_col(df, col, labels):
    
    edges = [df[col].describe()['min'],
             df[col].describe()['25%'],
             df[col].describe()['50%'],
             df[col].describe()['75%'],
             df[col].describe()['max']]
    df[col] = pd.cut(df[col], edges, labels = labels, duplicates = 'drop')
    return df

In [ ]:
labels = ['not_popular', 'below_avg', 'average', 'popular']

categorize_col(df, 'Vote_Average', labels)

df['Vote_Average'].unique()

In [ ]:
df.head()

In [ ]:
df['Vote_Average'].value_counts()

In [ ]:
df.dropna(inplace=True)
df.isna().sum()

In [ ]:
df.head()

# Phase 4 — Linear Regression Model

In [ ]:
df_ml = df.copy()
df_ml.head()

In [ ]:
df_ml.info()

## Feature Selection

In [ ]:
# Removing unnecessary columns

df_ml = df_ml.drop(columns=['Title'])

df_ml.head()

In [ ]:
## Checking Data Types
df_ml.dtypes

In [ ]:
## Encoding Vote Average
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()

df_ml['Vote_Average'] = le.fit_transform(df_ml['Vote_Average'])
df_ml.head()

In [ ]:
print(le.classes_)

## Final Feature Selection

In [ ]:
# Removing Genre column for the first model

df_ml = df_ml.drop(columns = ['Genre'])

df_ml.head()

# Phase 5 - Splitting Features and Targets

In [ ]:
# Features (Independent Variables)
X = df_ml.drop(columns=['Popularity'])

# Target (Dependent Variable)
y = df_ml['Popularity']

print("Features:")
display(X.head())

print('\nTarget:')
display(y.head())

In [ ]:
print('Shape of X :', X.shape)
print('Shape of y:', y.shape)

# Phase 6 — Train Test Split

In [ ]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size = 0.2,
    random_state = 42
)

In [ ]:
print('Training Features :', X_train.shape)
print('Testing Features :', X_test.shape)

print()

print('Training Target :', y_train.shape)
print('Testing Target :', y_test.shape)

print("Model trained successfully!")


# Phase 7 — Training the Linear Regression Model

In [ ]:
from sklearn.linear_model import LinearRegression

model = LinearRegression()

model.fit(X_train, y_train)

# Phase 8 — Making Predictions

In [ ]:
# Predicting popularity using test data

y_pred = model.predict(X_test)

print("First 10 predictions:\n")
print(y_pred[:10])

In [ ]:
prediction_df = X_test.copy()

prediction_df['Actual_Popularity'] = y_test
prediction_df['Predicted_Popularity'] = y_pred

prediction_df = prediction_df.round(2)
prediction_df.head(10)

# Phase 9 — Model Evaluation

In [ ]:
from sklearn.metrics import (
    mean_absolute_error,
    mean_squared_error,
    r2_score
)



In [ ]:
mae = mean_absolute_error(y_test, y_pred)

mse = mean_squared_error(y_test, y_pred)

rmse = np.sqrt(mse)

r2 = r2_score(y_test, y_pred)

In [ ]:
print(f"Mean Absolute Error: {mae:.2f}")
print(f"Mean Squared Error: {mse:.2f}")
print(f"Root Mean Squared Error: {rmse:.2f}")
print(f"R2 Score: {r2:.2f}")

In [ ]:
print(df.columns)
df_ml.columns

In [ ]:
df_ml_v2 = df.copy()

df_ml_v2 = df_ml_v2.drop(columns=['Title'])

df_ml_v2.head()

In [ ]:
df_ml_v2['Genre'].head()

## Encoding Genre column

In [ ]:
genre_encoded = df_ml_v2['Genre'].str.get_dummies(sep=', ')
genre_encoded.head(10)

In [ ]:
df_ml_v2 = pd.concat([df_ml_v2, genre_encoded], axis=1)

df_ml_v2 = df_ml_v2.drop(columns=['Genre'])

df_ml_v2.head()

In [ ]:
df_ml_v2.shape

# Phase 11 — Improved Linear Regression Model

In [ ]:
from sklearn.preprocessing import LabelEncoder

le_v2 = LabelEncoder()

df_ml_v2['Vote_Average'] = le_v2.fit_transform(df_ml_v2['Vote_Average'])

df_ml_v2.head()

In [ ]:
X_v2 = df_ml_v2.drop(columns=['Popularity'])

y_v2 = df_ml_v2['Popularity']

In [ ]:
from sklearn.model_selection import train_test_split

X_train_v2, X_test_v2, y_train_v2, y_test_v2 = train_test_split(
    X_v2,
    y_v2,
    test_size=0.2,
    random_state=42
)

In [ ]:
from sklearn.linear_model import LinearRegression

model_v2 = LinearRegression()

model_v2.fit(X_train_v2, y_train_v2)

In [ ]:
y_pred_v2 = model_v2.predict(X_test_v2)

In [ ]:
# Evaluation 
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
import numpy as np

mae_v2 = mean_absolute_error(y_test_v2, y_pred_v2)
mse_v2 = mean_squared_error(y_test_v2, y_pred_v2)
rmse_v2 = np.sqrt(mse_v2)
r2_v2 = r2_score(y_test_v2, y_pred_v2)

print(f"MAE  : {mae_v2:.2f}")
print(f"MSE  : {mse_v2:.2f}")
print(f"RMSE : {rmse_v2:.2f}")
print(f"R²   : {r2_v2:.4f}")

In [ ]:
df.describe()

In [ ]:
df[['Popularity', 'Vote_Count']].describe()

## Distribution Analysis

In [ ]:
plt.figure(figsize=(8,5))
plt.hist(df['Vote_Count'], bins=40,
         color='steelblue',
         edgecolor='black')
plt.title("Vote Count Distribution")
plt.xlabel("Vote Count")
plt.ylabel("Frequency")
plt.grid(alpha=0.3)
plt.show()

In [ ]:
plt.figure(figsize=(8,5))

plt.hist(df['Popularity'], bins=40,
         color='tomato',
         edgecolor='black')

plt.title("Popularity Distribution")
plt.xlabel("Popularity")
plt.ylabel("Frequency")

plt.grid(alpha=0.3)

plt.show()

## Log Transformation on Vote Count

In [ ]:
df_ml_v3 = df_ml_v2.copy()

df_ml_v3['Vote_Count'] = np.log1p(df_ml_v3['Vote_Count'])

df_ml_v3[['Vote_Count']].head()

In [ ]:
plt.figure(figsize=(8,5))

plt.hist(df_ml_v3['Vote_Count'],
         bins=40,
         color='green',
         edgecolor='black')

plt.title("Vote Count After Log Transformation")
plt.xlabel("Log(Vote Count)")
plt.ylabel("Frequency")

plt.grid(alpha=0.3)

plt.show()

In [ ]:
X_v3 = df_ml_v3.drop(columns=['Popularity'])
y_v3 = df_ml_v3['Popularity']

In [ ]:
from sklearn.model_selection import train_test_split

X_train_v3, X_test_v3, y_train_v3, y_test_v3 = train_test_split(
    X_v3,
    y_v3,
    test_size=0.2,
    random_state=42
)

In [ ]:
from sklearn.linear_model import LinearRegression

model_v3 = LinearRegression()
model_v3.fit(X_train_v3, y_train_v3)

In [ ]:
y_pred_v3 = model_v3.predict(X_test_v3)

In [ ]:
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
import numpy as np

mae_v3 = mean_absolute_error(y_test_v3, y_pred_v3)
mse_v3 = mean_squared_error(y_test_v3, y_pred_v3)
rmse_v3 = np.sqrt(mse_v3)
r2_v3 = r2_score(y_test_v3, y_pred_v3)

print(f"MAE  : {mae_v3:.2f}")
print(f"MSE  : {mse_v3:.2f}")
print(f"RMSE : {rmse_v3:.2f}")
print(f"R²   : {r2_v3:.4f}")

In [ ]:
df_ml_v3.dtypes

## Experiment 1 — Baseline Linear Regression

Features Used:
- Release_Date
- Vote_Count
- Vote_Average

Results:
- MAE  : 33.50
- MSE  : 24984.54
- RMSE : 158.06
- R²   : 0.0300

### Insight
The baseline model was trained using three numerical features. Although the model was able to make predictions, the R² score indicates that these features explain only a small portion of the variation in movie popularity.

## Experiment 2 — Genre Encoding

Changes Made:
- One-Hot Encoded the Genre column.
- Retained the original Vote_Count values.

Results:
- MAE  : 33.48
- MSE  : 24810.99
- RMSE : 157.52
- R²   : 0.0320

### Insight
Adding genre information provided a slight improvement in model performance. This suggests that genre contributes to movie popularity, although its impact is limited when used with a Linear Regression model.

## Experiment 3 — Log Transformation

Changes Made:
- Applied Log Transformation on Vote_Count.
- Used the Genre Encoded dataset.

Results:
- MAE  : 34.65
- MSE  : 24941.86
- RMSE : 157.93
- R²   : 0.0269

### Insight
Although the log transformation normalized the Vote_Count distribution, it did not improve the model performance. Therefore, this preprocessing step was not retained in the final model.

In [ ]:
comparison = pd.DataFrame({
    "Experiment": [
        "Baseline",
        "Genre Encoding",
        "Genre + Log Transform"
    ],
    "MAE": [33.50, 33.48, 34.65],
    "RMSE": [158.06, 157.52, 157.93],
    "R² Score": [0.0300, 0.0320, 0.0269]
})

comparison

# Conclusion

This project explored the use of Linear Regression to predict movie popularity using the available Netflix movie dataset.

Three different experiments were performed to improve the model:
1. Baseline Linear Regression
2. Genre Encoding
3. Log Transformation of Vote_Count

Among these, Genre Encoding produced the best results and was selected as the final model.

Despite multiple preprocessing experiments, the R² score remained relatively low. This indicates that movie popularity is influenced by additional factors that are not present in the dataset or that the relationship between the available features and popularity is not purely linear.

Future improvements may include:
- Collecting additional movie-related features.
- Performing more advanced feature engineering.
- Experimenting with non-linear machine learning models such as Decision Tree Regressor or Random Forest Regressor.

# Final Model Selection

After evaluating all the experiments, **Experiment 2 (Genre Encoding)** was selected as the final model because it achieved the best predictive performance among all Linear Regression models.

### Comparison of Experiments

| Experiment | MAE | RMSE | R² Score |
|------------|-----:|------:|---------:|
| Baseline Model | 33.50 | 158.06 | 0.0300 |
| Genre Encoding | **33.48** | **157.52** | **0.0320** |
| Genre + Log Transformation | 34.65 | 157.93 | 0.0269 |

**Selected Model:** Experiment 2 (Genre Encoding)

In [ ]:
# Final Selected Model

final_model = model_v2

X_train_final = X_train_v2
X_test_final = X_test_v2

y_train_final = y_train_v2
y_test_final = y_test_v2

y_pred_final = y_pred_v2

In [ ]:
print("Final Model Selected Successfully!")
print(final_model)

In [ ]:
print(f"Final R² Score : {r2_v2:.4f}")
print(f"Final RMSE     : {rmse_v2:.2f}")
print(f"Final MAE      : {mae_v2:.2f}")

# Final Model Visualization

In [ ]:
plt.figure(figsize=(8,6))

plt.scatter(
    y_test_final,
    y_pred_final,
    alpha=0.6,
    color='royalblue',
    edgecolors='black'
)

plt.plot(
    [y_test_final.min(), y_test_final.max()],
    [y_test_final.min(), y_test_final.max()],
    color='red',
    linestyle='--',
    linewidth=2,
    label='Ideal Prediction'
)

plt.title("Actual vs Predicted Popularity (Final Model)")
plt.xlabel("Actual Popularity")
plt.ylabel("Predicted Popularity")

plt.legend()

plt.grid(alpha=0.3)

plt.tight_layout()
plt.savefig("outputs/final_model_prediction.png", dpi=300, bbox_inches='tight')
plt.show()

# Final Conclusion

The objective of this project was to predict movie popularity using Linear Regression.

Three different experiments were performed to improve the model:

- Baseline Linear Regression
- Genre Encoding
- Log Transformation

Among these experiments, Genre Encoding achieved the highest R² score and was selected as the final model.

Although the final model successfully completed the prediction pipeline, the low R² score indicates that the available features are insufficient to accurately predict movie popularity using a linear relationship.

This project demonstrates the complete machine learning workflow, including:

- Data Cleaning
- Feature Engineering
- Encoding
- Model Training
- Model Evaluation
- Experimentation
- Model Selection

Future improvements may include:

- Collecting additional movie-related features
- Applying advanced feature engineering
- Using non-linear regression algorithms such as Decision Tree Regressor or Random Forest Regressor

## Final Prediction Results

The table below compares the actual popularity values with the values predicted by the final Linear Regression model.

In [ ]:
prediction_result = pd.DataFrame({
    "Actual Popularity": y_test_final,
    "Predicted Popularity": y_pred_final
})

prediction_result = prediction_result.round(2)

prediction_result.head(10)